ArduPlane/commands_logic.cpp

```diff
@@ -383,6 +390,58 @@ void Plane::do_nav_wp(const AP_Mission::Mission_Command& cmd)
     set_next_WP(cmd.content.location);
 }
 
+inline int modulo(int a, int b) {
+    if (b < 0) {
+        return modulo(-a, -b);
+    }
+    const int result = a % b;
+    return result >= 0 ? result : result + b;
+}
+
+void Plane::do_nav_into_wind(const AP_Mission::Mission_Command& cmd)
+{
+    Vector3f wind = ahrs.wind_estimate();
+    float wind_to = M_PI_2 - atan2f(wind.y, wind.x);
+    if (wind_to < 0.0) {
+        wind_to += M_2PI;
+    }
+
+    //get the next waypoint position
+    AP_Mission::Mission_Command next_cmd;
+    const uint16_t current_index = mission.get_current_nav_index() + 1;
+    mission.get_next_nav_cmd(current_index, next_cmd);
+
+    gcs_send_text_fmt(MAV_SEVERITY_WARNING, "Nav_into_wind: wind to %.1f", degrees(wind_to));
+
+    //verify the limits of the approach
+    if ( (cmd.content.wind.deg_start >= 360 && cmd.content.wind.deg_stop > 360) ||
+        (cmd.content.wind.deg_start >= cmd.content.wind.deg_stop) ) {
+        gcs_send_text(MAV_SEVERITY_WARNING, "Nav_into_wind: invalid limits, will not respect");
+    } else { //appply the limits (if needed)
+        //cartesian angle in radians to cardinal angle in degrees
+        uint16_t wind_from_deg = modulo((90 - degrees(wind_to)), 360);
+        if ( (wind_from_deg < cmd.content.wind.deg_start) || (wind_from_deg > cmd.content.wind.deg_stop) ) {
+            //choose the nearest angle inside the limits
+            uint16_t diff_low = 180.0 - fabs(fmod(fabs(wind_from_deg - cmd.content.wind.deg_start), 360.0) - 180.0);
+            uint16_t diff_high = 180.0 - fabs(fmod(fabs(wind_from_deg - cmd.content.wind.deg_stop), 360.0) - 180.0);
+            if (diff_low < diff_high) {
+                wind_to = radians(modulo((90 - (cmd.content.wind.deg_start)), 360));
+            } else {
+                wind_to = radians(modulo((90 - (cmd.content.wind.deg_stop)), 360));
+            }
+        }
+    }
+
+    //set a waypoint away from the wind, radius away and at altitude over the next wp
+    Location next_cmd_loc = next_cmd.content.location;
```

Comment:
Sanity check to make sure that the next_cmd is actually some form of a navigational waypoint, otherwise if its a do_command the location read out of storage will be garbage.